### ISSUE TITLE
Configure a pipeline via API

### ISSUE THREAD

**Issue by zabil** [gocd/gocd#1417]
Related to #1404 

Build a new API to save pipeline configuration. 

PR => #1491 

**GET**:  `curl -i 'http://<go-server>:8153/go/api/admin/pipelines/<pipeline-name>' -u 'user:password' -H 'Accept: application/vnd.go.cd.v1+json' -H 'Content-Type: application/json'  -H "If-None-Match: <Etag>"`

**PUT**: `curl -i 'http://<go-server>:8153/go/api/admin/pipelines/<pipeline-name>' -u 'user:password' -H 'Accept: application/vnd.go.cd.v1+json' -H 'Content-Type: application/json' -H 'If-Match: <Etag>' -X PUT -d '{<pipeline json>}'`

**POST**: `curl -i 'http://<go-server>:8153/go/api/admin/pipelines' -u 'user:password' -H 'Accept: application/vnd.go.cd.v1+json' -H 'Content-Type: application/json' -X POST -d '{"group": "<group-name>", "pipeline": <pipeline json>}'`

Here's a draft structure without HAL:

```JSON
{
  "name": "yourproject",
  "label_template": "foo-1.0.${COUNT}-${svn}",
  "locked": true,
  "template_name": null,
  "params": [
    {
      "name": "COMMAND",
      "value": "echo"
    },
    {
      "name": "WORKING_DIR",
      "value": "/repo/branch"
    }
  ],
  "tracking_tool": {
    "type": "mingle",
    "attributes": {
      "base_url": "http://mingle.example.com",
      "project_identifier": "my_project",
      "grouping_conditions": "status > 'In Dev'"
    }
  },
  "timer": "0 0 22 ? * MON-FRI",
  "materials": [
    {
      "type": "svn",
      "attributes": {
        "name": "svn_material",
        "url": "http://your-svn/",
        "dest": "dest_folder",
        "auto_update": true
      }
    }
  ],
  "environment_variables": [
    {
      "name": "MULTIPLE_LINES",
      "value": "multiplelines",
      "secure": true
    },
    {
      "name": "COMPLEX",
      "value": "This has very <complex> data",
      "secure": false
    }
  ],
  "stages": [
    {
      "name": "ut",
      "fetch_materials": true,
      "clean_working_directory": false,
      "never_clean_artifacts": true,
      "approval": {
        "type": "manual|success",
        "authorization": {
          "roles": [],
          "users": []
        }
      },
      "environment_variables": [
        {
          "name": "MULTIPLE_LINES",
          "value": "multiplelines",
          "secure": true
        },
        {
          "name": "COMPLEX",
          "value": "This has very <complex> data",
          "secure": false
        }
      ],
      "jobs": [
        {
          "name": "build",
          "run_on_all_agents": false,
          "run_instance_count": 0,
          "timeout": null,
          "environment_variables": [],
          "resources": [
            "jdk5",
            "tomcat5"
          ],
          "tasks": [
            {
              "type": "ant",
              "attributes": {
                "target": "clean",
                "working_dir": "dir"
              }
            }
          ],
          "artifacts": [
            {
              "src": "target/dist.jar",
              "dest": "pkg",
              "type": "build"
            },
            {
              "src": "target/reports/**/*Test.xml",
              "dest": "reports",
              "type": "test"
            }
          ],
          "tabs": [
            {
              "name": "coverage",
              "path": "Jcoverage/index.html"
            }
          ],
          "properties": [
            {
              "name": "coverage.class",
              "src": "target/emma/coverage.xml",
              "xpath": "substring-before(//report/data/all/coverage[starts-with(@type,'class')]/@value, '%')"
            }
          ]
        }
      ]
    }
  ]
}
```

# Checklist of Items to be done

* [x] Dependency material validation that upstream exists
* [x] Fetch artifact validation that upstream/ancestor exists
* [x] Support for etags - #1509
* [x] Rename stage/job - validate if downstream pipelines are affected
* [x] Response of PUT API to include all validation errors against the fields
* [x] circular dependency check
* [x] do not allow pipeline name change
* [ ] validations when template is associated with a pipeline and stages are being posted
* [ ] api to create package repo and package definition as pipeline api allows only reference to an existing pkg repo/def - #1507 
* [ ] api to create pluggable scm material as pipeline api allows only reference to an existing pluggable scm - #1508 
* [ ] currently xsd validations are being invoked before save, should we skip that for this api?
* [ ] material type and task type need to be specific string values. API should generate descriptive error message when an unexpected type is provided.

**Comment by ketan** [comment]
@all - We were thinking about whether it makes sense to rename some attributes and break away from the names used in the cruise-config.xml file.

Specifically we think that the `isLocked` attribute on a pipeline does not correctly describe what the flag does. The name of the attribute feels like it is describing the state of the pipeline. Just like `running` or `building` or `scheduled`.

We were thinking about renaming it to `enable_pipeline_locking` or `run_multiple_pipelines` or something along these lines to describe what the flag does.

Does anyone have better ideas on what we could call this?

Are there other attributes on other config elements that you think need some thought?

**Comment by arvindsv** [comment]
I think "enable_pipeline_locking" makes sense. Consider that pipeline locking enables two things: "run single instance" and "lock pipeline on failure". It is a long-standing request to separate the two through different flags. That's why I think "enable_pipeline_locking" is more appropriate.

I wonder if we should think about having a top-level item called "triggers" or something, which encompasses both "timer" and "materials". But, I realize that "triggers" is not a term we use.

**Comment by scottmuc** [comment]
I'm with @ketan and don't think the configuration should be too much of an abstraction from the xml config. Also, the config XML is already so clean, wouldn't it be reasonable to just use it as is?

I like that I can fetch config with a script like this `fetch-pipeline-config` (and I have a similar one to select on the pipeline group):

```
#!/usr/bin/env bash

set -e

go_server=$1
pipeline_name=$2
rm -f pipeline.xml
curl -s -o /tmp/pipelines.xml http://${go_server}:8153/go/api/admin/config/current.xml
xmllint --xpath "//pipeline[@name='${pipeline_name}']" --format /tmp/pipelines.xml > pipeline.xml
rm -f /tmp/pipelines.xml
echo $(pwd)/pipeline.xml
```

I would love to have a something to push this config and get config validation from the API.

Working with pipeline abstractions in jenkins has a huge impedance mismatch.

Then again, this JSON looks pretty close to the XML. I'll be ready to test this on a real system as soon as there's an RC for this!

**Comment by zabil** [comment]
There are a few more reasons why we need this JSON API.

* Support easy UI development. 
* [JSON HAL](http://stateless.co/hal_specification.html) support so that you can link to all the resources on the pipeline config in the future.

Although not related to format I think using JSON will make it cleaner/easier to implement   
* API versioning 
* Support for [PATCH](http://tools.ietf.org/html/rfc5789) for easier updates from curl
* Pessimistic locking and conflict resolution using etags.  

There's also a long discussion at #838 where we talk about moving away from exposing our XML configuration.

**Comment by scottmuc** [comment]
Cool, thanks for the context @zabil!

**Comment by arvindsv** [comment]
I was thinking about this. Suppose we use this to create a pipeline, then there will be other information that needs to be passed — say, the pipeline group that this pipeline needs to be created in.

Does that mean that this JSON will go in as a part of the payload, with that ancillary information? Something like:

```
PUT /pipeline

{
  "group": "abc",
  ...
  "payload": {
    "name" ...
  }
}
```

Or something else?

**Comment by ketan** [comment]
@arvindsv - A "PUT" implies you're updating an existing pipeline, in which case a pipeline group is not necessary.

Where you will need a pipeline group is when doing a "POST", in which case, the "POST", should happen at a different url.

```
POST /go/api/pipeline_groups/:pipeline_group_name
Content-Type: application/json

"pipeline": {
  "name": "DeployToProduction"
}
```

The response would be -

```
201 Created
Location: /go/api/admin/pipelines/:pipeline_name
```

Any subsequent PUTs need to happen at the url returned by POST.

**Comment by arvindsv** [comment]
Sorry, I meant to say POST. What you say makes sense. Thank you! Just to take it a little further: In my example, there was only one piece of ancillary information (pipeline group). What if there was more? Say group name, position of pipeline in group and say something else. What would change?

**Comment by ketan** [comment]
> Just to take it a little further: In my example, there was only one piece of ancillary information (pipeline group). What if there was more? Say group name, position of pipeline in group and say something else. What would change?

As things stand right now, "order" is not a property of the pipeline, but that of the pipeline group. I suppose you could do one of two things here - add a separate "re-order" operation (move pipeline X before/after pipeline Y) — OR — do what we're doing with the new pipeline API and allow submitting the entire pipeline group object with re-ordered pipelines (may kill performance).

**Comment by arvindsv** [comment]
> do what we're doing with the new pipeline API and allow submitting the entire pipeline group object with re-ordered pipelines (may kill performance).

Yes, it probably will. We're really doing that?? I'd just add it to the end, and reorder later too.

**Comment by drmuey** [comment]
not sure if this belongs here or not but I was pointed here in https://groups.google.com/forum/#!topic/go-cd/cvx7HML3C6A so I wanted to be sure it was either covered here or would get its own issue.

It'd be immensley helpful is the pipeline scheduler API could return the counter of the build. Its OK to queue it and wait on stuff to actually start but having the build ID (AKA the pipeline 'counter') immediately is invaluable. Check out the thread above for some of the hacky ways its being done unreliably ATM :)

Thanks!

**Comment by ketan** [comment]
@arvindsv, @scottmuc @tomzo - About the JSON structure described above. I am not sure if we're all agreeing on the same thing :)

Here's what we'll be doing. The JSON format will be freshly designed. This means the JSON structure will not be an AS-IS port from XML, there will be some places where the JSON structure deviates from the XML structure in order to add some consistency if needed.

In the long run we hope that users move to using API's as a means to configure pipelines over editing XML directly.

**Comment by tomzo** [comment]
But whatever new model you will create, it will still have to be transformed to old XML right? You cannot go too far with deviations. 
As I mentioned before, I'd like to adjust my extension point to be closer to what you will come up with rather than old XML. Let me know when there is something more or less complete.

**Comment by ketan** [comment]
@tomzo - we're thinking of transforming the old XML to closely reflect JSON (eventually), to fix some consistency issues we have in mind.

Would that be a problem wrt the work you're doing?

**Comment by arvindsv** [comment]
> not sure if this belongs here or not but I was pointed here in https://groups.google.com/forum/#!topic/go-cd/cvx7HML3C6A so I wanted to be sure it was either covered here or would get its own issue.

@drmuey: No, this is about config API of the pipeline. The issue you want is #990, which you've already commented in. As I said there, it's not easy to do, given the way it is being done now. It is possible, though. If you or someone else wants to contribute, I can help show you around and give you an idea.

**Comment by tomzo** [comment]
I have already implemented a full JSON model which very close to current xml. Then I created all conversion from that JSON to config-api classes. 

> we're thinking of transforming the old XML to closely reflect JSON (eventually)

Currently xml is tied to the config-api classes. What do plan do about that? You cannot have differences between JSON and XML when they are both tied to config-api.

By *eventually* you mean "not in this iteration"? Xml stays the same when you release the single-page app?

> Would that be a problem wrt the work you're doing?

I don't think so. It depends on how far will be drift away from current model. Do you plan to rewrite the config-api types or just the xml schema?

**Comment by ketan** [comment]
We're not going to change the config-api classes. Just the XML schema.

**Comment by arvindsv** [comment]
If the schema changes (say a renamed attribute), won't the config-api classes change?

**Comment by ketan** [comment]
> If the schema changes (say a renamed attribute), won't the config-api classes change?

They would, but that is not something we'd be doing as part of this change, ideally the attribute names on models should be de-coupled from the attribute names in XML.

**Comment by drmuey** [comment]
@arvindsv thanks, I thought it seemed wrong :) I'll post in #990 then because I have an idea that would be very simple and, I beleive, solve all the problems and I'd love to send up a pull request if I can take you up on your offer to show me around ;) thanks

**Comment by tomzo** [comment]
> They would, but that is not something we'd be doing as part of this change, ideally the attribute names on models should be de-coupled from the attribute names in XML

Please note that this is the only thing you can do as part of decoupling xml from config-api as long as there is reflection-based serialization in xml.
If you will ever want to change the OO model then it would force change in xml schema

**Comment by rajiesh** [comment]
Tested Config API on 15.3.0-2663 and issues identified are logged on issue #1630

**Comment by rajiesh** [comment]
verified on 15.3.0-2771(81951df8c0ea7ccc13382160aa4defd5c816a206) and defects raised in #1630 are all fixed and validations passed
